docs/proposals/20201123-tracing-ingestion-api-and-opentelemetry-collector.md

```diff
@@ -0,0 +1,87 @@
+Table of contents generated with markdown-toc
+</a>
+
+</i>
+
+</small>
+
+## TLDR
+
+This document proposes to expose tracing ingestion API  backed by the OpenTelemetry collector. This is a first step to support ingestion and storage for tracing data. The OpenTelemetry collector gives flexibility to forward tracing data to variety of tracing platforms, both on-prem or SaaS - Jaeger, Zipkin, Tempo, SigNoz or DataDog, Dynatrace etc.
+
+## Why
+
+At the moment Observatorium does not expose API to ingest distributed tracing data.
+
+## Goals
+
+* Expose API in the Observatorium API service to ingest tracing data.
+* Deploy a component that receives tracing data and is able to forward it to tracing platforms with storage.
+
+## Non-Goals
+
+* Store data in a persistent storage.
+* Provide read API for tracing data.
+
+## How
+
+The how is split into two sections:
+1. Expose tracing ingestion API
+2. Add deployment manifests for OpenTelemetry collector
+
+### Expose tracing ingestion API in the Observatorium API service
+
+There are a couple of open-source tracing protocols out there - Zipkin, Jaeger and OpenTelemetry. At the moment the OpenTelemetry seems to be the [most popular](https://opentelemetry.io/vendors/), and it is projected to have the biggest adoption in the future. Hence, using OpenTelemetry protocol seems to be the most appropriate choice.
+
+The OpenTelemetry protocol primarily supports gRPC for sending traces (OTLP gRPC) with proto encoding. The HTTP protocol with protobuf encoding is supported as well, but the JSON encoding is still in experimental mode. The majority of users use OTLP gRPC and the HTTP is used in environments where gRPC cannot be used (e.g. mobile clients). Because the Observatorium already supports HTTP and the final state is to support both OTLP HTTP and gRPC, the Observatorium could initially support HTTP and in parallel start working on gRPC.
+
+#### Multitenancy
+
+Multitenancy could be handled the same way as it is handled for metrics and logs. The API service could add an HTTP header or attribute/label to data. This label would then be used in the collector to identify the tenant. Then the collector could route the data to a specific exporter per tenant or export the data to platform/storage by using a single exporter.
+
+### Add deployments manifests for OpenTelemetry collector
+
+The OpenTelemetry collector community already support deploying the collector via [OpenTelemetry operator](https://github.com/open-telemetry/opentelemetry-operator) or [HELM chart](https://github.com/open-telemetry/opentelemetry-helm-charts). A plain Kubernetes manifests could be as well used given the stateless nature and low complexity of the collector.
+
+## Action Plan
+
+* Iterate and finalise this design document.
+* Add deployment manifests for OpenTelemetry collector.
+* Add OTLP HTTP with binary encoding to the Observatorium API.
+* Add OTLP gRPC support to the Observatorium API.
```

Comment:
Thanos has a gRPC store API, and I believe the Loki Push API also works with gRPC.
- Is the intention to have a single gRPC endpoint for all three types, or one endpoint for OTLP, one for Loki Push, and one for Prometheus Remote Write
- Is it too early to request that the Observatorium gRPC endpoint support introspection?

Reply:
> Thanos has a gRPC store API, and I believe the Loki Push API also works with gRPC. Yet those endpoints are not exposed on Observatorium. Observatorium should support either none or eventually all of the protocols via gRPC.

Good point, perhaps these gRPC endpoints are just implementation/internal details of mentioned systems. 
If we decide to expose multiple gRPC services then there are various ways how it can be done (single vs multiple ports). 

> Is it too early to request that the Observatorium gRPC endpoint support introspection?

This feature should be discussed separately and not in this proposal.

Reply:
> Thanos has a gRPC store API, and I believe the Loki Push API also works with gRPC. Yet those endpoints are not exposed on Observatorium. Observatorium should support either none or eventually all of the protocols via gRPC.

Where this assumption comes from? Thanos storeAPI is an internal API which is never exposed. API is for external users, and it just happen that HTTP APIs are the most popular (Prometheus Query APIs, remote write and so on). 

I see no problem in adding gRPC OTLP if we need, even if other APIs are HTTP. Let's do it if we see grpc otlp more popular. Just a bit more work to adjust our middleware for interceptors if we want them 👍🏽

Reply:
> Is it too early to request that the Observatorium gRPC endpoint support introspection?

No problem with https://github.com/grpc/grpc/blob/master/doc/server-reflection.md if that is what we want.